File: backend/app/sources/binance.py
"""
Binance Futures data source.

Public, keyless endpoints. We pull only what's needed for the
intelligence engine: klines, 24h ticker, open interest,
funding, long/short ratio, recent forced liquidations (top trader stream
is paid; we approximate using long/short and OI deltas).
"""
from __future__ import annotations

import asyncio
import logging
from typing import Any, Dict, List, Optional

import httpx

log = logging.getLogger(__name__)

BASE_FUTURES = "https://fapi.binance.com"


class BinanceClient:
    def __init__(self, timeout: float = 8.0) -> None:
        self._client = httpx.AsyncClient(
            base_url=BASE_FUTURES,
            timeout=timeout,
            headers={"User-Agent": "cit-terminal/0.1"},
        )

    async def close(self) -> None:
        await self._client.aclose()

    async def _get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:
        try:
            r = await self._client.get(path, params=params)
            r.raise_for_status()
            return r.json()
        except Exception as e:  # noqa: BLE001
            log.warning("binance %s failed: %s", path, e)
            return None

    async def klines(self, symbol: str, interval: str = "1m", limit: int = 200) -> List[List[Any]]:
        data = await self._get("/fapi/v1/klines",
                               {"symbol": symbol, "interval": interval, "limit": limit})
        return data or []

    async def ticker_24h(self, symbol: str) -> Dict[str, Any]:
        return await self._get("/fapi/v1/ticker/24hr", {"symbol": symbol}) or {}

    async def open_interest(self, symbol: str) -> Dict[str, Any]:
        return await self._get("/fapi/v1/openInterest", {"symbol": symbol}) or {}

    async def open_interest_hist(self, symbol: str, period: str = "5m", limit: int = 30) -> List[Dict[str, Any]]:
        return await self._get(
            "/futures/data/openInterestHist",
            {"symbol": symbol, "period": period, "limit": limit},
        ) or []

    async def funding_rate(self, symbol: str) -> Dict[str, Any]:
        data = await self._get("/fapi/v1/premiumIndex", {"symbol": symbol})
        return data or {}

    async def long_short_ratio(self, symbol: str, period: str = "5m") -> List[Dict[str, Any]]:
        return await self._get(
            "/futures/data/globalLongShortAccountRatio",
            {"symbol": symbol, "period": period, "limit": 1},
        ) or []

    async def top_trader_long_short(self, symbol: str, period: str = "5m") -> List[Dict[str, Any]]:
        return await self._get(
            "/futures/data/topLongShortPositionRatio",
            {"symbol": symbol, "period": period, "limit": 1},
        ) or []

    async def taker_buysell_volume(self, symbol: str, period: str = "5m") -> List[Dict[str, Any]]:
        return await self._get(
            "/futures/data/takerlongshortRatio",
            {"symbol": symbol, "period": period, "limit": 1},
        ) or []

    async def snapshot(self, symbol: str) -> Dict[str, Any]:
        """Pull everything we need for one symbol concurrently."""
        (klines, t24, oi, oi_hist, funding,
         lsr, top_lsr, taker) = await asyncio.gather(
            self.klines(symbol, "1m", 200),
            self.ticker_24h(symbol),
            self.open_interest(symbol),
            self.open_interest_hist(symbol, "5m", 30),
            self.funding_rate(symbol),
            self.long_short_ratio(symbol),
            self.top_trader_long_short(symbol),
            self.taker_buysell_volume(symbol),
        )
        return {
            "klines": klines,
            "ticker24h": t24,
            "openInterest": oi,
            "openInterestHist": oi_hist,
            "funding": funding,
            "longShortRatio": lsr,
            "topTraderRatio": top_lsr,
            "takerRatio": taker,
        }

File: frontend/src/server/binance.ts
/**
 * Binance Futures source - public, keyless endpoints.
 * Server-side only. Uses native fetch (Node 18+ runtime).
 */

const BASE = "https://fapi.binance.com";

async function get<T>(path: string, params: Record<string, any> = {}): Promise<T | null> {
  const qs = new URLSearchParams(
    Object.entries(params).reduce<Record<string, string>>((a, [k, v]) => {
      if (v !== undefined && v !== null) a[k] = String(v);
      return a;
    }, {})
  ).toString();
  const url = `${BASE}${path}${qs ? `?${qs}` : ""}`;
  try {
    const r = await fetch(url, {
      headers: { "User-Agent": "cit-terminal/0.1" },
      // Cache for 4s server-side - cheap rate limit safety net.
      next: { revalidate: 4 },
    });
    if (!r.ok) return null;
    return (await r.json()) as T;
  } catch {
    return null;
  }
}

export interface BinanceSnapshot {
  klines: any[][];
  ticker24h: any;
  openInterest: any;
  openInterestHist: any[];
  funding: any;
  longShortRatio: any[];
  topTraderRatio: any[];
  takerRatio: any[];
}

export async function fetchSnapshot(symbol: string): Promise<BinanceSnapshot> {
  const [
    klines, ticker24h, openInterest, openInterestHist, funding,
    longShortRatio, topTraderRatio, takerRatio,
  ] = await Promise.all([
    get<any[][]>("/fapi/v1/klines", { symbol, interval: "1m", limit: 200 }),
    get<any>("/fapi/v1/ticker/24hr", { symbol }),
    get<any>("/fapi/v1/openInterest", { symbol }),
    get<any[]>("/futures/data/openInterestHist", { symbol, period: "5m", limit: 30 }),
    get<any>("/fapi/v1/premiumIndex", { symbol }),
    get<any[]>("/futures/data/globalLongShortAccountRatio", { symbol, period: "5m", limit: 1 }),
    get<any[]>("/futures/data/topLongShortPositionRatio", { symbol, period: "5m", limit: 1 }),
    get<any[]>("/futures/data/takerlongshortRatio", { symbol, period: "5m", limit: 1 }),
  ]);

  return {
    klines: klines ?? [],
    ticker24h: ticker24h ?? {},
    openInterest: openInterest ?? {},
    openInterestHist: openInterestHist ?? [],
    funding: funding ?? {},
    longShortRatio: longShortRatio ?? [],
    topTraderRatio: topTraderRatio ?? [],
    takerRatio: takerRatio ?? [],
  };
}
